Add p-ext support with spec v0.94

linsinan1995 opened the pull request:
Add p-ext support with spec v0.93

* Andes-tech and Shao-Chung Wang implemented the RISC-V 'p' extension with the spec version 0.51 with Binutils 2.32
* Sinan Lin updated the implementation to the spec version 0.93 with Binutils 2.36

linsinan1995 committed:
Add p-ext support with spec v0.94

Nelson1225 commented on gas/config/tc-riscv.c:
```diff
@@ -1083,6 +1118,7 @@ validate_riscv_insn (const struct riscv_opcode *opc, int length)
       case 'P': USE_BITS (OP_MASK_PRED, OP_SH_PRED); break;
       case 'Q': USE_BITS (OP_MASK_SUCC, OP_SH_SUCC); break;
       case 'o': /* ITYPE immediate, load displacement.  */
+      case 'l': /* IMM6L */
```
This looks strange.  We encode it as ENCODE_ITYPE_IMM here (validate_riscv_insn), but will use ENCODE_SBTYPE_IMM6L to encode it later in the riscv_ip.

linsinan1995 replied:
Thanks for pointing it out. Fixed it the latest commit.

Nelson1225 commented on gas/config/tc-riscv.c:
```diff
@@ -2410,6 +2492,91 @@ riscv_ip (char *str, struct riscv_cl_insn *ip, expressionS *imm_expr,
+		      {
+			if (xlen == 32 && (regno % 2) != 0)
+			  {
+			    as_bad (_("The number of Rs1 must be even "
```
It would be great if the error/warning messages are starting with lower case letter, and without the period at the end.

linsinan1995 replied:
Thanks for pointing it out. Fixed it the latest commit.

Nelson1225 commented on gas/testsuite/gas/riscv/insn-dsp.d:
```diff
@@ -0,0 +1,245 @@
+#as: -mext-dsp
```
This looks weird, I don't see any implementation about the option in this PR.

linsinan1995 replied:
I updated all test cases in the latest commit. Thanks for spotting it out!

Nelson1225 commented on include/opcode/riscv.h:
```diff
@@ -63,6 +63,8 @@ static const char * const riscv_pred_succ[16] =
 
 #define EXTRACT_ITYPE_IMM(x) \
   (RV_X(x, 20, 12) | (RV_IMM_SIGN(x) << 12))
+#define EXTRACT_ITYPE_IMM6L(x) \
```
ITYPE or SBTYPE?  Is this a typo?

linsinan1995 replied:
Thanks. Fixed it the latest commit.

Nelson1225 commented on gas/testsuite/gas/riscv/insn-dsp64.d:
```diff
@@ -0,0 +1,91 @@
+
+0+000 <dsp64>:
+[ 	]+.*:[ 	]+.*[ 	]+add32[ 	]+r1,r2,r3
+.*R_RISCV_RELAX_ENTRY.*
```
We don't have this relocation.  If you need new relocations, you should file an issue to riscv-psabi first, we need the approvals of riscv community, and developers of gnu and llvm.

linsinan1995 replied:
Test cases are updated in the latest commit. Thanks for spotting it out!

Nelson1225 reviewed (commented):
I think you probably miss the porting from riscv-dis.c, since you add several of new operands, but you don't teach the dis-assembler to recognize these operands.  Therefore, I suppose the gas test cases will be failed.

linsinan1995 commented on gas/config/tc-riscv.c:
```diff
@@ -973,6 +985,29 @@ arg_lookup (char **s, const char *const *array, size_t size, unsigned *regnop)
+#define MAX_KEYWORD_LEN 32
+
+static bfd_boolean
+parse_nds_v5_field (const char **str, char name[MAX_KEYWORD_LEN])
```
‘nds’ was the prefix before spec 0.80. Since we are using the old prefix `nds`  as a prefix so I keep the 'nds'. I will change it to `rvp`. Thanks.

linsinan1995 commented on gas/config/tc-riscv.c:
```diff
@@ -1118,6 +1154,40 @@ validate_riscv_insn (const struct riscv_opcode *opc, int length)
 	     return FALSE;
 	  }
 	break;
+      case 'n':
```
I checked 'p', 'P', 'r'(for rvp), but they are taken by others. so I used the old prefix 'nds'. It would be great to have a discussion on that. Thanks!

kito-cheng commented:
`nds` means abbreviation of `andes`, which used in Andestech internal only, P extension will become standard extension rather than Andestech specific extension,  so I think `nds` should gone in this implementation.

linsinan1995 commented:
Hi @kito-cheng , I did not find a proper prefix for p ext operands, so keep 'nds' as prefix (since old spec also used it as a prefix for intrinsic functions). Prefixes like 'p', 'P', 'r' (rvp) and 'd' (dsp) are taken. I just filed an [issue] (https://github.com/riscv/riscv-binutils-gdb/issues/258) for this problem. Thanks.

Nelson1225 commented:
I shouldn't have time to check the encoding, does the encoding conflicts are resolved in the v0.93 spec?  If they are resolved, and the previous comments are also fixed, then the patches looks good to me.  The operand naming doesn't really matter, just renamed to rps or others are fine to me, just avoid using the vendor prefixes since p is a standard extension.  Thanks.

linsinan1995 commented:
> I shouldn't have time to check the encoding, does the encoding conflicts are resolved in the v0.93 spec? If they are resolved, and the previous comments are also fixed, then the patches looks good to me. The operand naming doesn't really matter, just renamed to rps or others are fine to me, just avoid using the vendor prefixes since p is a standard extension. Thanks.

I will check those encoding conflicts and then inform you. Thanks.

pz9115 committed:
Fix disassmeble kmada32 to it alais name kmar64

linsinan1995 committed:
Sync version flag with GCC

Nelson1225 commented on bfd/elfxx-riscv.c:
```diff
@@ -1770,6 +1771,15 @@ riscv_parse_add_implicit_subsets (riscv_parse_subset_t *rps)
+  if (riscv_lookup_subset (rps->subset_list, "p", &subset))
+    {
+      riscv_parse_add_subset (rps, "zpn",
+                              RISCV_UNKNOWN_VERSION,
```
Wrong GNU indents.

Nelson1225 commented on gas/config/tc-riscv.c:
```diff
@@ -131,6 +131,8 @@ static const struct riscv_ext_version ext_version_table[] =
   {"c", ISA_SPEC_CLASS_20190608, 2, 0},
   {"c", ISA_SPEC_CLASS_2P2,      2, 0},
 
+  {"p", ISA_SPEC_CLASS_DRAFT, 2, 0},
```
Same as the g and k expansions, but we don't need to care about this here.

Nelson1225 commented on gas/config/tc-riscv.c:
```diff
@@ -973,6 +985,29 @@ arg_lookup (char **s, const char *const *array, size_t size, unsigned *regnop)
+    *p++ = *str_t++;
+  *p = '\0';
+
+  if (strncmp (name, "nds_", 4) == 0)
```
This is minor, so it is OK here.  But once it becomes to one of the stable standard extensions, then I think we should give them new operand names.

Nelson1225 reviewed (approved):
Same GNU coding standard Indents, it would great to fix them if you plan to go upstream.  Otherwise looks good to me.  Thanks.

linsinan1995 committed:
set swap16 and smbb32 alias to pkbt16 and mulsr64

linsinan1995 commented:
Some update on the PR

- squash fixup commits
- fix wrong indents in `bfd/elfxx-riscv.c`
- change swap16 and smbb32 to be an alias of pkbt16 and mulsr64

hope51607 commented on include/opcode/riscv-opc.h:
```diff
@@ -621,6 +621,667 @@
+#define MASK_SRLI16_U 0xff00707f
+#define MATCH_STAS16 0xf4002077
+#define MASK_STAS16 0xfe00707f
+#define MATCH_STSA16 0x46003077
```
Hi, according to [riscv-p-spec](https://github.com/riscv/riscv-p-spec/blob/master/P-ext-proposal.adoc#stsa16-simd-16-bit-straight-subtraction-addition) and [riscv-opcodes](https://github.com/riscv/riscv-opcodes/blob/master/opcodes-rvp#L185), I think MATCH_STSA16 should be 0xf6002077.

linsinan1995 replied:
Fixed. thanks!

marcfedorow commented on gas/testsuite/gas/riscv/insn-dsp.s:
```diff
@@ -0,0 +1,298 @@
+	kabs16    a1, a2
+	clrs16    a1, a2
+	clz16     a1, a2
+	clo16     a1, a2
```
swap16?

marcfedorow commented on gas/testsuite/gas/riscv/insn-dsp.s:
```diff
@@ -0,0 +1,298 @@
+	uclip8    a1, a2, 3
+	clrs8     a1, a2
+	clz8      a1, a2
+	clo8      a1, a2
```
swap8?

marcfedorow commented on include/opcode/riscv.h:
```diff
@@ -229,6 +253,8 @@ static const char * const riscv_pred_succ[16] =
 #define OP_MASK_RL		0x1
 #define OP_SH_RL		25
+#define OP_MASK_RC		0x1f
+#define OP_SH_RC		27
```
OP_MASK_RS3

marcfedorow commented on opcodes/riscv-opc.c:
```diff
@@ -828,6 +828,339 @@ const struct riscv_opcode riscv_opcodes[] =
+{"ave",         0, INSN_CLASS_ZPN, "d,s,t",     MATCH_AVE, MASK_AVE, match_opcode, 0 },
+{"bitrev",      0, INSN_CLASS_ZPN, "d,s,t",     MATCH_BITREV, MASK_BITREV, match_opcode, 0 },
+{"bitrevi",     0, INSN_CLASS_ZPN, "d,s,l",     MATCH_BITREVI, MASK_BITREVI, match_opcode, 0 },
+{"bpick",       0, INSN_CLASS_ZPN, "d,s,t,nds_rc", MATCH_BPICK, MASK_BPICK, match_opcode, 0 },
```
nds_rc => r

linsinan1995 replied:
It looks to me that `nds_rc` can be replaced by `r` with the patch.
```
diff --git a/opcodes/riscv-dis.c b/opcodes/riscv-dis.c
index e6b5cba205..f20dc3db64 100644
--- a/opcodes/riscv-dis.c
+++ b/opcodes/riscv-dis.c
@@ -306,6 +306,11 @@ print_insn_args (const char *d, insn_t l, bfd_vma pc, disassemble_info *info)
          print (info->stream, "%s", riscv_gpr_names[rs1]);
          break;
 
+       case 'r':
+         print (info->stream, "%s",
+                riscv_gpr_names[EXTRACT_OPERAND (RS3, l)]);
+         break;
+
```

The regression test passes, but I am not sure about whether this change would affect the behaviour of .insn directive. Hi @Nelson1225 , any suggestion on it?

marcfedorow commented on bfd/elfxx-riscv.c:
```diff
@@ -1770,6 +1771,15 @@ riscv_parse_add_implicit_subsets (riscv_parse_subset_t *rps)
+      riscv_parse_add_subset (rps, "zpsf",
+			      RISCV_UNKNOWN_VERSION,
+			      RISCV_UNKNOWN_VERSION, TRUE);
+    }
```
"zprv"?

linsinan1995 committed:
fix wrong format of swap16 & add testcases

linsinan1995 commented on opcodes/riscv-opc.c:
```diff
@@ -1025,7 +1025,7 @@ const struct riscv_opcode riscv_opcodes[] =
 {"sunpkd832",   0, INSN_CLASS_ZPN, "d,s",       MATCH_SUNPKD832, MASK_SUNPKD832, match_opcode, 0 },
 {"swap8",       0, INSN_CLASS_ZPN, "d,s",       MATCH_SWAP8, MASK_SWAP8, match_opcode, 0 },
-{"swap16",      0, INSN_CLASS_ZPN, "d,s",       MATCH_SWAP16, MASK_SWAP16, match_opcode, 0 },
+{"swap16",      0, INSN_CLASS_ZPN, "d,s,t",       MATCH_PKBT16, MASK_PKBT16, match_rs1_eq_rs2, INSN_ALIAS },
```
totally agree. thanks a lot for catching up! I have made a fix on this.

linsinan1995 committed:
update to 0.96: remove zprv

linsinan1995 committed:
update to 0.96: remove clo*

linsinan1995 committed:
remove rv64 add/sub 64

linsinan1995 committed:
update testcase to spec 0.9.10

linsinan1995 committed:
update to spec 0.9.10

linsinan1995 committed:
disassemble rs3 argument ('r' modifiler)

linsinan1995 committed:
remove minw, maxw and bpick

linsinan1995 committed:
fix rdov and clrov and remove persudo insn in DECLARE_INSN

linsinan1995 commented:
the new commit fixes an error that csr* instructions can't access RVP csr, and adds a test case.
